Add callback functions to runBlock

bmoneke opened the pull request:
Callback
---------

The runBlock method allows to specify a callback function, which I'd like to use instead of polling the picoscope whether it has finished acquiring samples.

A change in picobase.runBlock is required, but it does not break existing code (after having changed the parameters of _lowLevelRunBlock in all ps-xxxx files):

For this the low level function and the function in picobase have to be changed. Changing the low level function is easy. Giving None as the callback function works as well, instead of an empty void pointer.
Changing the function in picobase would require all low level functions to accept this parameter. I checked all ps-xxxx files and all library calls (except ps2000) have this parameter, such that I'd add the callback as a parameter to all low level functions.
The picobase method could have None as default parameter for the low level function (see code), such that this is handed to the low levels which would do nothing different than today.

The callback has to be a c-function. In order to turn any python function into a c-function, I added decorators (for runBlock and transferData) wrapping the function.
Right now these wrappers are in ps4000a, but maybe they could be in picobase, as they should be the same for every scope.

I tested the function and it works. You can emit a Qt signal or do something else, whenever the picosope finishes its block.


Docstrings
----------
I changed some docstings, in order to use the possibilities of restruct text and expand them. Some missing docstrings I added.


Questions
----------
- What are your thoughts regarding the callback function?
- I'd like to improve the docstrings, would you appreciate it?

colinoflynn commented:
For me adding the callback makes sense - but @hmaarrfk is the real developer, I haven't been active on this for some time, so would wait for his feedback too. However would welcome the documentation updates as well alongside this, any documentation improvements are of course appreciated too. Thanks for helping out!

hmaarrfk commented:
I just want to make sure that you actually tested all the additions you made with real hardware.

If you do, then I am more inclined to push them. If you haven't I would rather you constrain your changes to the change you actually tested.

Specifically, I am looking at the ps5000 and ps6000.

I can open an issue pointing to your PR, but I don't want to have a hidden untested code path that crashes people stuff.

bmoneke commented:
I did not test my changes with any other hardware than a ps4824 (my only model), but the change should not break anything, because I add just two parameters to _lowLevelRunBlock: callback, pParameter.
Whenever the manual of the specific model shows the same c-library call (psXXXXaRunBlock) description as the one of ps4000A, then I handed these parameters to that library call as well. All manuals, except of ps2000, ps3000 (without A), have the same library call description.
I understand your hesitation. For that reason I checked all manuals.

If you prefer, I can just keep the added parameters and revert the changes to the library call.
The parameter change is necessary in order to make picobase and the psXXXX files work together. Otherwise, I'd have to add a second runBlock method to picobase.

bmoneke commented:
I reverted the changes to the API calls. Now the only changes to other picoscopes is, that the '_lowLevelRunBlock' function accepts more parameters, which are not used at all, except in the ps4000a.
This is necessary, because the baseclass hands these parameters to the picoscopes.

hmaarrfk commented:
Hmm. Ok. so the API is consistent between all of them. Thanks. I can review later this week. Just ping me.

bmoneke commented:
Yes, the API is consistent between all the lowLevels, only ps4000a got real changes.

I'll have access to the 5000a and 6000 series in the near future, such that I could add and test the callback functionality with them as well. You can wait with merging until I have changed and tested the code, or I'll make a new pull request in the future with just their code.

If you prefer for ease of reviewing, that I split this pull request in two, one for the expanded comments (no code change) and one for the actual callback functionality, I would do that.

hmaarrfk commented on picoscope/ps4000a.py:
```diff
@@ -59,14 +59,60 @@
+    """
+    callback = CFUNCTYPE(c_void_p, c_int16, c_int32, c_uint32, c_int16,
+                         c_uint32, c_int16, c_int16, c_void_p)
+    return callback
```
```suggestion
    return callback(function)
```

hmaarrfk commented on picoscope/ps5000.py:
```diff
@@ -264,13 +264,16 @@ def _lowLevelSetSimpleTrigger(self, enabled, trigsrc, threshold_adc,
             c_int16(oversample), byref(timeIndisposedMs),
             c_uint32(segmentIndex), c_void_p(), c_void_p())
+        # According to the documentation, 'callback, pParameter' should work
+        # instead of the last two c_void_p parameters.
```
```suggestion
        # instead of the last two c_void_p parameters.
        # However to avoid the potential for serious crashes, we decided to
        # not include them in this function call.
```

hmaarrfk commented on picoscope/ps5000a.py:
```diff
@@ -274,14 +287,15 @@ def _lowLevelSetSimpleTrigger(self, enabled, trigsrc, threshold_adc,
             c_uint32(numPostTrigSamples), c_uint32(timebase),
             byref(timeIndisposedMs), c_uint32(segmentIndex),
-            c_void_p(), c_void_p())
+            callback, pParameter)
```
```suggestion
            blockReady(callback), pParameter)
```

Is this not more in the spirit of these low level py files?

bmoneke replied:
You're right, that would be better. It should work that way as he decorotator doesn't do it any different. I'll try it.

bmoneke replied:
Using blockReady(callback) instead of the decorator does not work.

The reason might be, that according to https://docs.python.org/3/library/ctypes.html#callback-functions the ctype function has to remain in memory the whole time. Which is the case, if the function is decorated. But in the case of 'blockReady(callback)' it is not stored as a ctype function.

bmoneke commented:
Now it works. Thanks for hinting that way of "storage".

In this commit I also removed the pParameter.
